### PR TITLE
feat(block-embed): block-embed and component refactor #584

### PR DESCRIPTION
**Pull request by arkRedM**

### CODE REVIEW

**Comment by arkRedM** [comment]
I tried to reason about this patch below

- https://www.loom.com/share/030ba64e3c8144be805e04e6abc3dcf6
- https://www.loom.com/share/f0fded2cafc04ef6a28553414d03c542

Lint is failing at specter-recursive-paths(probably no way around it?)
Unsure why style is failing. Something to do with spaces?

File: src/cljs/athens/effects.cljs [hunk 1]
```diff
@@ -303,10 +303,13 @@
...
         (.blur active-el))
       (js/setTimeout (fn []
-                       (let [html-id (str "#editable-uid-" uid)
+                       (let [[uid embed-id]  (db/uid-and-embed-id uid)
+                             html-id         (str "editable-uid-" uid)
                              ;;targets (js/document.querySelectorAll html-id)
                              ;;n       (count (array-seq targets))
-                             el      (js/document.querySelector html-id)]
+                             el              (cond
+                                               embed-id (js/document.querySelector (str "textarea[id^=" html-id "-embed-]"))
```
Comment (tangjeff0) [inline src/cljs/athens/effects.cljs:311]: Wow, I actually didn't know this was possible. So what exactly is happening here?

File: src/cljs/athens/events.cljs [hunk 2]
```diff
@@ -704,36 +701,40 @@
...
                            (empty? value)
                            [:unindent uid d-key-down context-root-uid]
 
                            (and (zero? start) value)
-                           [:enter/bump-up uid])]
-    {:dispatch event}))
+                           [:enter/bump-up uid new-uid])]
+    {:dispatch-later [{:ms 0  :dispatch event}
+                      {:ms 10 :dispatch [:editing/uid (cond-> new-uid
+                                                        embed-id (str "-embed-" embed-id))]}]}))
```
Comment (tangjeff0) [inline src/cljs/athens/events.cljs:737]: Incredible refactor.

File: src/cljs/athens/views/blocks.cljs [hunk 3]
```diff
@@ -834,8 +841,13 @@
...
                         :linked-ref/open (or (false? linked-ref) initial-open)})]
 
-     (fn [block linked-ref-data]
+     (fn [block linked-ref-data opts]
        (let [{:block/keys [uid string open children _refs]} block
+             uid-sanitized-block (s/transform
+                                   (specter-recursive-path #(contains? % :block/uid))
+                                   (fn [{:block/keys [original-uid uid] :as block}]
+                                     (assoc block :block/uid (or original-uid uid)))
+                                   block)
```
Comment (tangjeff0) [inline src/cljs/athens/views/blocks.cljs:850]: Reminder to learn this

**Review by tangjeff0 — CHANGES_REQUESTED** [review]
Lots of comments, and I need to take my own time to learn some of the techniques you are using. But the fundamentals here are absolutely amazing. Well done @arkRedM. This is incredible.

File: src/cljs/athens/components.cljs [hunk 4]
```diff
@@ -16,45 +21,79 @@
...
-  [content _]
-  [:button content])
-
-
-(defn render-component
-  "Renders a component using its parse tree & its uid."
+(defn span-click-stop
+  []
+  [:span {:on-click (fn [e]
+                      (.. e stopPropagation))}])
```
[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/components.cljs]: added a comment in code as well
"Stop clicks from propagating to textarea and thus preventing edit mode"

File: src/cljs/athens/views/node_page.cljs [hunk 5]
```diff
@@ -363,25 +363,29 @@
...
                                             (swap! state assoc :block new-B :parents new-P)))}
                [parse-and-render (or title string) (:block/uid block)]]))]
-         [block-el block linked-ref-data]]))))
+         [block-el
+          (recursively-modify-block-for-embed block embed-id)
+          linked-ref-data
+          {:block-embed? true}]]))))
 
 
 (defn linked-ref-el
```
[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/views/node_page.cljs:392]: width - increased
drag and up/down working
enter behavior creates child

[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/views/node_page.cljs:392]: width too small fixed
textarea focus issue is fixed
enter on top level block now creates child
empty block enter creates new child
shift up/down + drag and drop issues fixed

File: src/cljs/athens/util.cljs [hunk 6]
```diff
@@ -2,17 +2,37 @@
...
+(defn recursively-modify-block-for-embed
+  "Modify the block and all the block children to have same embed-id for
+   referencing the embed block rather than block in original page"
+  [block embed-id]
+  (s/transform
+    (specter-recursive-path #(contains? % :block/uid))
+    (fn [{:block/keys [uid] :as block}]
+      (assoc block :block/uid (str uid "-embed-" embed-id)
+                   :block/original-uid uid))
+    block))
```
[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/util.cljs:33]: https://www.youtube.com/watch?v=VTCy_DkAJGk -- this will take it from 0-100 very quick

https://docs.devarkm.com/clojure/specter -- my notes and this is for final reference of how powerful it can be and how easy it is to extend it

File: src/cljs/athens/db.cljs [hunk 7]
```diff
@@ -219,6 +219,14 @@
        (= uid)))
 
 
+(defn uid-and-embed-id
+  [uid]
+  (or (some->> uid
+               (re-find #"^(.+)-embed-(.+)")
+               rest vec)
+      [uid nil]))
```
[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/db.cljs:227]: yeah. saved a ton of time. no big refactor and leverages already built code

File: src/cljs/athens/effects.cljs (hunk 1 above)
Comment (arkRedM) [reply src/cljs/athens/effects.cljs]: updated it to new and added comments 
querySelector can have a regex(ish -- very limited) search https://stackoverflow.com/questions/8714090/queryselector-wildcard-element-match

File: src/cljs/athens/views/blocks.cljs (hunk 3 above)
Comment (arkRedM) [reply src/cljs/athens/views/blocks.cljs:851]: I love specter for this exact reason. It's almost impossible to write this code in plain clj(s) -- and it will be very hard to read.

File: src/cljs/athens/db.cljs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tangjeff0) [reply src/cljs/athens/db.cljs:227]: This reminded me that Roam also uses uuids in their elements, even though block/uids already exist. Perhaps Roam is doing something similar to achieve ephemeral uniqueness, in the case where multiple blocks with the same uid can be displayed. 

image

Other interesting things in this screenshot too of their HTML data tags

**Review by tangjeff0 — CHANGES_REQUESTED** [review]
https://www.loom.com/share/604668befa0e445eb4c20d312f7d0ff2

File: src/cljs/athens/keybindings.cljs [hunk 8]
```diff
@@ -88,7 +88,8 @@
    ["Tomorrow"      mui-icons/Today (fn [] (str "[[" (:title (get-day -1)) "]]")) nil nil]
    ["Yesterday"     mui-icons/Today (fn [] (str "[[" (:title (get-day 1)) "]]")) nil nil]
    ["YouTube Embed" mui-icons/YouTube "{{[[youtube]]: }}" nil 2]
-   ["iframe Embed"  mui-icons/DesktopWindows "{{iframe: }}" nil 2]])
+   ["iframe Embed"  mui-icons/DesktopWindows "{{iframe: }}" nil 2]
+   ["Block Embed"   mui-icons/ViewDayRounded "{{[[embed]]: (())}}" nil 4]])
```
[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/keybindings.cljs:92]: A ton of edge cases. But I think it's best to catch these edge cases early because they are easy to fix if there is context. Else it will be hard to pick them up completely new at some other point. Keep adding more cases where you think it can start failing other than the usual suspects.   

Regarding parens -- is it an arbitrary number of matching parens(>2)? -- change should be simple enough

[the remark this answers is not in the document]
Comment (tangjeff0) [reply src/cljs/athens/keybindings.cljs:92]: Seems arbitrary

image

source (feel free to add tests): https://roamresearch.com/#/app/athensresearch/page/kWDwfUWfL

Another case while creating the Roam demo above: block embed in linked refs? Maybe it'll work itself out.

[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/keybindings.cljs:92]: This patch is already bloated. Let me take up the parser stuff(this and #668) in the next patch

[the remark this answers is not in the document]
Comment (arkRedM) [reply src/cljs/athens/keybindings.cljs:92]: test and behavior decisions

https://www.loom.com/share/a66a953aa915449299954734c7c5a8a5
https://www.loom.com/share/8117e03734ca4c8ba10c49399944b575

**Review by tangjeff0 — CHANGES_REQUESTED** [review]
3 more edge cases

- indent/unindent
- shift up/down (and also click and drag, forgot to mention in video)
- delete

https://www.loom.com/share/2b105e6228054388a25ee924ab870c92

**Comment by arkRedM** [comment]
test for these edge cases: https://www.loom.com/share/6565d4e56e7e4bc389ce5c24feddee9d

Regarding the focus, I think Intelli j and most editors do a good job of focusing when navigating up/down. I also think focus logic is too distributed and we can probably combine them.

**Review by tangjeff0 — COMMENTED** [review]
> I also think focus logic is too distributed and we can probably combine them.

Agree, this can be done better.

Two changes requested

https://www.loom.com/share/004b4fc02cd24968bea8edc001d9b04e

- indent when you press enter on an empty block leads to unexpected behavior
	- normal block: should refocus on right caret position
	- block embed / linked ref: should unindent for blocks that are more than one level indented
- block embed shouldn't show anything for an invalid block uid, e.g. `{{[[embed]]: ((invalid))}}`. My code below.


```
(defmethod component :block-embed
  [content uid]
  ;; bindings are eval only once in with-let
  ;; which is needed to keep embed integrity else it will update on
  ;; each re-render. Similar to ref-comp
  (let [block-embed-uid (last (re-find #"\(\((.+)\)\)" content))]
    ;; not reactive. some cases where delete then ctrl-z doesn't work
    (if (db/e-by-av :block/uid block-embed-uid)
      (r/with-let [embed-id (random-uuid)]
        [:div.block-embed (use-style block-embed-adjustments)
         (let [block (db/get-block-document [:block/uid block-embed-uid])]
           [:<>
            [blocks/block-el
             (recursively-modify-block-for-embed block embed-id)
             {:linked-ref false}
             {:block-embed? true}]
            (when-not @(subscribe [:editing/is-editing uid])
              [:> mui-icons/Edit
               {:on-click (fn [e]
                            (.. e stopPropagation)
                            (dispatch [:editing/uid uid]))}])])])
      ;; roam actually hides the brackets around [[embed]]
      [:span "{{" content "}}"])))

```

**Comment by arkRedM** [comment]
https://www.loom.com/share/cf4068bdfced47689e1d3367c8bf79c5 
unindent and invalid blocks

**Review by tangjeff0 — APPROVED** [review]
